splice_machine/src/main/java/com/splicemachine/derby/impl/sql/execute/operations/RecursiveUnionOperation.java

```diff
@@ -0,0 +1,120 @@
+import com.splicemachine.derby.stream.iapi.DataSet;
+import com.splicemachine.derby.stream.iapi.DataSetProcessor;
+import com.splicemachine.derby.stream.iapi.OperationContext;
+import com.splicemachine.utils.Pair;
+import org.apache.log4j.Logger;
+import org.spark_project.guava.base.Strings;
+
+/**
+ * Created by yxia on 3/22/19.
+ */
+public class RecursiveUnionOperation extends UnionOperation {
+    private static Logger LOG = Logger.getLogger(RecursiveUnionOperation.class);
+    private DataSet rightDS;
+    protected static final String NAME = RecursiveUnionOperation.class.getSimpleName().replaceAll("Operation","");
+
+    @Override
+    public String getName() {
+        return NAME;
+    }
+
+    public RecursiveUnionOperation() {
+        super();
+    }
+
+    public RecursiveUnionOperation(SpliceOperation leftResultSet,
+                          SpliceOperation rightResultSet,
+                          Activation activation,
+                          int resultSetNumber,
+                          double optimizerEstimatedRowCount,
+                          double optimizerEstimatedCost) throws StandardException{
+        super(leftResultSet, rightResultSet, activation, resultSetNumber, optimizerEstimatedRowCount, optimizerEstimatedCost);
+    }
+
+    @Override
+    public String toString() {
+        return "RecursiveUnionOperation{" +
+                "left=" + leftResultSet +
+                ", right=" + rightResultSet +
+                '}';
+    }
+
+    @Override
+    public String prettyPrint(int indentLevel) {
+        String indent = "\n"+ Strings.repeat("\t",indentLevel);
+
+        return "RecursiveUnion:" + indent + "resultSetNumber:" + resultSetNumber
+                + indent + "leftResultSet:" + leftResultSet
+                + indent + "rightResultSet:" + rightResultSet;
+    }
+
+    @Override
+    public DataSet<ExecRow> getDataSet(DataSetProcessor dsp) throws StandardException {
+        if (!isOpen)
+            throw new IllegalStateException("Operation is not open");
+
+        OperationContext operationContext = dsp.createOperationContext(this);
+        operationContext.pushScope();
+
+        // get the loop limit from configuration
+        SConfiguration configuration= EngineDriver.driver().getConfiguration();
+        int MAX_LOOP = configuration.getRecursiveQueryIterationLimit();
+
+        // compute the seed
+        // For control path, the dataset returned by materialize can be read only once as the iterator cannot be reset,
+        // so we have to return two datasets
+        Pair<Pair<DataSet, DataSet>, Integer> right = leftResultSet.getDataSet(dsp).persistIt2();
+        DataSet resultDS = right.getFirst().getFirst();
+        rightDS = right.getFirst().getSecond();
+
+        // compute the recursive body
+        right = rightResultSet.getDataSet(dsp).persistIt2();
+
+        int loop = 0;
+        while (right.getSecond() > 0 && loop < MAX_LOOP) {
```

Comment:
Other DBs raise an exception if the limit is reached, I think we should do the same. As someone else suggested it would be useful for MAX_LOOP to be a session level property.

Reply:
recursion limit can now be controlled by a session level property recursivequeryiterationlimit.
Also runtime exception will be raised when limit is reached.